homeassistant/components/cover/template.py

```diff
@@ -0,0 +1,353 @@
+    SUPPORT_SET_TILT_POSITION, SUPPORT_OPEN, SUPPORT_CLOSE, SUPPORT_STOP,
+    SUPPORT_SET_POSITION, ATTR_POSITION, ATTR_TILT_POSITION)
+from homeassistant.const import (
+    CONF_FRIENDLY_NAME, CONF_ENTITY_ID,
+    EVENT_HOMEASSISTANT_START, MATCH_ALL,
+    CONF_VALUE_TEMPLATE, CONF_ICON_TEMPLATE,
+    STATE_OPEN, STATE_CLOSED)
+from homeassistant.exceptions import TemplateError
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import async_generate_entity_id
+from homeassistant.helpers.event import async_track_state_change
+from homeassistant.helpers.restore_state import async_get_last_state
+from homeassistant.helpers.script import Script
+
+_LOGGER = logging.getLogger(__name__)
+_VALID_STATES = [STATE_OPEN, STATE_CLOSED, 'true', 'false']
+
+CONF_COVERS = 'covers'
+
+CONF_POSITION_TEMPLATE = 'position_template'
+CONF_TILT_TEMPLATE = 'tilt_template'
+OPEN_ACTION = 'open_cover'
+CLOSE_ACTION = 'close_cover'
+STOP_ACTION = 'stop_cover'
+POSITION_ACTION = 'set_cover_position'
+TILT_ACTION = 'set_cover_tilt_position'
+
+TILT_FEATURES = (SUPPORT_OPEN_TILT | SUPPORT_CLOSE_TILT | SUPPORT_STOP_TILT |
+                 SUPPORT_SET_TILT_POSITION)
+
+COVER_SCHEMA = vol.Schema({
+    vol.Optional(CONF_VALUE_TEMPLATE): cv.template,
+    vol.Required(OPEN_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Required(CLOSE_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Required(STOP_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_POSITION_TEMPLATE): cv.template,
+    vol.Optional(CONF_TILT_TEMPLATE): cv.template,
+    vol.Optional(CONF_ICON_TEMPLATE): cv.template,
+    vol.Optional(POSITION_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(TILT_ACTION): cv.SCRIPT_SCHEMA,
+    vol.Optional(CONF_FRIENDLY_NAME, default=None): cv.string,
+    vol.Optional(CONF_ENTITY_ID): cv.entity_ids
+})
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_COVERS): vol.Schema({cv.slug: COVER_SCHEMA}),
+})
+
+
+@asyncio.coroutine
+def async_setup_platform(hass, config, async_add_devices, discovery_info=None):
+    """Set up the Template cover."""
+    covers = []
+
+    for device, device_config in config[CONF_COVERS].items():
+        friendly_name = device_config.get(CONF_FRIENDLY_NAME, device)
+        state_template = device_config.get(CONF_VALUE_TEMPLATE)
+        position_template = device_config.get(CONF_POSITION_TEMPLATE)
+        tilt_template = device_config.get(CONF_TILT_TEMPLATE)
+        icon_template = device_config.get(CONF_ICON_TEMPLATE)
+        open_action = device_config[OPEN_ACTION]
+        close_action = device_config[CLOSE_ACTION]
+        stop_action = device_config[STOP_ACTION]
+        position_action = device_config.get(POSITION_ACTION)
+        tilt_action = device_config.get(TILT_ACTION)
+
+        if ((position_template is None and state_template is None) or
+                (position_template is not None and
+                 state_template is not None)):
+            _LOGGER.error('Must specify only one of: %s',
```

Comment:
Instead of testing like this, use the voluptuous `Exclusive` validator to make sure only 1 is passed in. Example: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/wink.py#L53-L56